lib/rubber/cloud/datto.rb

```diff
@@ -0,0 +1,97 @@
+require 'rubber/cloud/generic'
+require 'httparty'
+
+module Rubber
+  module Cloud
+    class Datto < Base
+      def initialize(env, capistrano)
+        super(env, capistrano)
+      end
+
+      def endpoint
+        env.endpoint
+      end
+
+      def active_state
+        'running'
+      end
+
+      def stopped_state
+        'stopped'
+      end
+
+      def create_instance(instance_alias, image_name, image_type, security_groups, availability_zone, datacenter)
+        response = HttpAdapter.post(uriBuilder)
+        if response.code != 200
+          raise errFromResponse(response)
+        end
+
+        response["id"]
+      end
+
+      def describe_instances(instance_id=nil)
+        instances = []
+        if instance_id.nil?
+          response = HttpAdapter.get(uriBuilder)
+          if response.code != 200
+            raise errFromResponse(response)
+          end
+
+          response["results"].each do |item|
+            instances << instance_from_response(item)
+          end
+        else
+          response = HttpAdapter.get(uriBuilder(instance_id))
+          if response.code != 200
+            raise errFromResponse(response)
+          end
+
+          instances << instance_from_response(response)
+        end
+        instances
+      end
+
+      def destroy_instance(instance_id)
+        response = HttpAdapter.delete(uriBuilder(instance_id))
+        if response.code != 200
+          raise errFromResponse(response)
+        end
+        response
+      end
+
+      private
+
+      # @param [Hash] response - response hash representing an instance
+      # @param [Hash] converted hash conforming to the rubber instance api.
+      def instance_from_response(response)
+          instance = {}
+          instance[:id] = response["id"]
+          instance[:external_ip] = response["ip"]
+          instance[:internal_ip] = response["ip"]
+          instance[:state] = (!!response["workerRunning"]) ? active_state : stopped_state
+          instance[:provider] = 'datto'
+          instance[:platform] = Rubber::Platforms::LINUX
+          instance
+      end
```

Comment:
The body of this method is indented too far.

I think the form you propose is preferable if more complex logic is expected in this method later... Though in that case there might also be an argument for separating more complex logic out into instance methods, but I digress. More minimal syntax works for me too:

```ruby
def instance_from_response(response)
  {
    :id => response["id"],
    :external_ip => response["ip"],
    :internal_ip => response["ip"],
    :state => response["workerRunning"] ? active_state : stopped_state,
    :provider => "datto",
    :platform => Rubber::Platforms::LINUX,
  }
end
```